lib/linked_list.rb

```diff
@@ -19,99 +19,228 @@ def initialize
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+      @head = Node.new(value, @head)
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
+      checker = @head
+
+      until checker.nil?
+        return true if checker.data == value
+        checker = checker.next
+      end
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      return nil if @head.nil?
+      max = @head.data
+      marker = @head.next
+      until marker.nil?
+        max = marker.data if marker.data > max
+        marker = marker.next
+      end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+      min = @head.data
+      marker = @head.next
+      until marker.nil?
+        min = marker.data if marker.data < min
+        marker = marker.next
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
     def length
-      raise NotImplementedError
+      length = 0
+      current = @head    
+      until current.nil?
+        length += 1
+        current = current.next
+      end
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if index > self.length || index < 0
+      curr = @head
+
+      index.times do
+        curr = curr.next
+      end
+
+      return curr.data
     end
 
     # method to print all the values in the linked list
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      curr = @head
+
+      until curr.nil?
+        p curr.data
+        curr = curr.next
+      end
     end
 
     # method to delete the first node found with specified value
     def delete(value)
-      raise NotImplementedError
+      curr = @head
+      previous = nil
+
+      return nil if curr.nil?
+
+      until curr.data == value

```

Comment:
This could crash if `curr` is nil.

```suggestion
      until curr.nil?
```

Then inside the loop if you find the item to delete, make the deletion and return.